User space function metrics through bpf

kylebartush-twitter opened the pull request:
Problem

It is difficult to instrument some of the services written in c that
my team manages. We can't easily export metrics and are stuck
parsing logs to expose metrics. This causes lots of issues when
it comes to logging throttling by syslog and log rotation.

Solution

I've added the ability to configure the probing of user space
libraries through bpf. This will allow us to export metrics for
the number of times a specific function is called for a particular
shared library.

Result

I've added a sampler called usercall, that can be configured to probe
user space libraries and export metrics. This can be done through
the configuration file, either through a more friendly search heuristic
or through specifying the exact file and function you would like to be
probed.

kylebartush-twitter committed:
Initial plumbing for libcall sampler

kylebartush-twitter committed:
Update libcall to accept library->funcs in config

kylebartush-twitter committed:
Remove percentiles, add lib search paths

kylebartush-twitter committed:
Adding some lib searching logic and tests

kylebartush-twitter committed:
First run at actual bpf program.

This is completely untested, due to working on a mac. The major parts
should be connected though.

kylebartush-twitter committed:
Update so that non-called user functions export 0

kylebartush-twitter committed:
Allow for configuring specific library files

kylebartush-twitter committed:
Add some documentatino to the libcall config, lint fixes

kylebartush-twitter committed:
Bubble up errors rather than unwrapping

kylebartush-twitter committed:
Derive default rather than implementing

kylebartush-twitter committed:
Adding type aliases for config data structures

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/rezolus?pullRequest=221) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/rezolus?pullRequest=221) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/rezolus?pullRequest=221) it.</sub>

brayniac commented on configs/libcall.toml:
```diff
@@ -0,0 +1,67 @@
+# This example configuration covers detailed configuration for each sampler
```
Maybe change this text to talk about this particular sampler you're showing off?

kylebartush-twitter replied:
I shouldn't have left that boilerplate in there. I think the sample config covers most of this so I just cleaned things up a bit.

brayniac commented on src/samplers/libcall/config.rs:
```diff
@@ -0,0 +1,75 @@
+use serde_derive::Deserialize;
```
License text should be included at the top of each file. You can grab it from one of the existing files and just set it to 2021 to reflect the year the initial code was written.

kylebartush-twitter replied:
Added

brayniac commented on src/samplers/libcall/bph.c:
```diff
@@ -0,0 +1 @@
+// TODO
```
Should this file be in the commit?

kylebartush-twitter replied:
Nice catch, I never ended up using this. Removed now.

brayniac commented on src/samplers/libcall/mod.rs:
```diff
@@ -0,0 +1,278 @@
+use async_trait::async_trait;
```
License header

kylebartush-twitter replied:
Added

brayniac commented on src/samplers/libcall/mod.rs:
```diff
@@ -0,0 +1,278 @@
+use std::path::Path;
+
+#[allow(dead_code)]
+pub struct LibCall {
```
I think you might want to rename this to `Libcall` or consider some other name entirely for both the module and sampler structs?

Typically, we're using the convention that if we use `FooBar` that the config section and module would be called `foo_bar`. You seem to be using `LibCall` and `libcall` instead - which implies it's a single word and should therefore have `Libcall` as its title-case name.

kylebartush-twitter replied:
We discussed this a bit and I've renamed everything from LibCall -> Usercall.

brayniac commented on src/samplers/libcall/stat.rs:
```diff
@@ -0,0 +1,24 @@
+impl LibCallStatistic {
+    pub fn new(library: &str, func: &str) -> Self {
+        Self {
+            stat_path: format!("{}/{}", library, func),
```
I'm wondering if we should prefix these at all - I'm trying to imagine how this would look mixed in with host-level stats that are namespaced by component. If the intent is to run just the single sampler as a service-specific sampler it's maybe less necessary to have a namespace prefix.

kylebartush-twitter replied:
I've added a namespace prefix: 'usercall' to all the metrics created by this sampler.

brayniac commented on src/samplers/libcall/stat.rs:
```diff
@@ -0,0 +1,24 @@
+use rustcommon_metrics::{AtomicU32, AtomicU64, Source, Statistic};
```
License header

kylebartush-twitter replied:
Added

brayniac commented on src/samplers/tcp/bpf.c:
```diff
@@ -85,4 +85,4 @@ int trace_tcp_rcv_state_process(struct pt_regs *ctx, struct sock *skp)
 
     start.delete(&skp);
     return 0;
-}
```
Agree with adding a newline here, but I'd maybe keep it out of this PR.

kylebartush-twitter replied:
I think this was an auto format that snuck in when I was looking at some of the other files for examples. I've reverted it.

brayniac commented:
Overall this PR looks pretty good. Have some minor changes that are needed before merge. Biggest thing I'm left wondering about is the sampler naming and if these stats should be namespaced or not.

Thanks for submitting the PR, this is a cool use-case for Rezolus and I'm excited to get this merged.

kylebartush-twitter committed:
Update libcall sample config comments

kylebartush-twitter committed:
Remove un-needed file

thinkingfish commented on src/samplers/libcall/mod.rs:
```diff
@@ -0,0 +1,278 @@
+        }
+
+        // Add probes by searching the default paths.
+        let default_paths: Vec<String> = vec![
```
Should the default path variable be promoted to the top of the module as a constant?

kylebartush-twitter replied:
Yeah, I think that's a good idea. I moved these paths to a const.

brayniac commented on src/samplers/libcall/mod.rs:
```diff
@@ -0,0 +1,278 @@
+}
+
+#[allow(dead_code)]
+fn path_match(lib_name: &str, path: &Path) -> bool {
```
What's the reason for doing this with string manipulation instead of using the Path/PathBuf types?

kylebartush-twitter replied:
The extension returned in the case of a lib of the form: libkrb.so.5.9.2.  is '2'.

brayniac replied:
See new feedback about maybe adding a comment about this for the function

kylebartush-twitter committed:
Adding license text to source

kylebartush-twitter committed:
Revert unrelated file change

kylebartush-twitter committed:
Make default paths a const

kylebartush-twitter committed:
Rename LibCall -> Usercall

kylebartush-twitter committed:
Add usercall namespace && run cargo fmt

brayniac commented on src/samplers/usercall/config.rs:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2019-2021 Twitter, Inc.
```
Please use just `2021` we have decided against using year ranges. (I may need to cleanup existing files)

brayniac commented on src/samplers/usercall/mod.rs:
```diff
@@ -0,0 +1,283 @@
+// Copyright 2019-2021 Twitter, Inc.
```
Same here for the year

brayniac commented on src/samplers/usercall/stat.rs:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2019-2021 Twitter, Inc.
```
Fix

brayniac commented on src/samplers/usercall/mod.rs:
```diff
@@ -0,0 +1,283 @@
+}
+
+#[allow(dead_code)]
+fn path_match(lib_name: &str, path: &Path) -> bool {
```
Might be worth adding a comment along the lines of your response to my asking about why this is string manipulation. It's not immediately obvious from the code, so worth putting a hint here as to why we need to do it this way.

brayniac commented on configs/usercall.toml:
```diff
@@ -0,0 +1,34 @@
+# versions of libkrb.
+#
+# WARNING: You cannot attach multiple probes to the same function.
+[samplers.usercall.lib_files.krb3]
```
For the config - I'm wondering if it might make sense to do:
```
[samplers.usercall]
enabled = true

[[samplers.usercall.libraries]]
name = "foo"
paths = [ "/path/to/libfoo.so" ]
functions = [ "bar", "baz" ]

[[samplers.usercall.libraries]]
name = "qux"
functions = ["qux_init" ]

[[samplers.usercall.libraries]]
name = "grault5"
paths = [ "/path/to/libgrault.so.5" ]
functions = [ "garply" ]

[[samplers.usercall.libraries]]
name = "grault3"
paths = [ "/path/to/libgrault.so.5" ]
functions = [ "garply" ]

[[samplers.usercall.libraries]]
name = "grault"
paths = [ "/path/to/libgrault.so.5", ""/path/to/libgrault.so.3" ]
functions = [ "garply" ]
```

This might be a little more verbose in the config, but it might offer us better flexibility if we want to add additional keys in the future - currently we're treating the keys as paths, which may limit what we can do in the future without a breaking change.

I also think you could use this to differentiate between what's currently 'lib_files' vs 'lib_search' by conditioning on if there are paths specified within the library table.

kylebartush-twitter replied:
The only issue I see with this configuration is that we may have different functions in different versions of the libs. 

We could do it like you suggested above and just drop the vector of paths. We lose the concept of 'combined' probes, but probably we can do that combination during querying. I guess there are pluses and minuses to both. I'm leaning slightly towards what you suggested with a single path and just getting rid of combined probes. Happy to write up whichever you prefer.

kylebartush-twitter replied:
Okay, got some updates in that should get the config to be of the format we discussed.

kylebartush-twitter committed:
Fix copyright comment and add explanation to path_match

kylebartush-twitter committed:
Update config format

brayniac commented on src/samplers/mod.rs:
```diff
@@ -99,6 +101,7 @@ pub trait Sampler: Sized + Send {
     /// Register all the statistics
     fn register(&self) {
         for statistic in self.sampler_config().statistics() {
+            info!("Registering statistic");
```
This isn't a super useful log message - including the statistic name and marking it debug would be more useful. Either that or remove it entirely. Does not seem like it should be info level.

kylebartush-twitter replied:
I added it to help me understand the code flow, but should have been removed.

kylebartush-twitter committed:
Remove unhelpful log line

brayniac commented on src/samplers/usercall/mod.rs:
```diff
@@ -0,0 +1,285 @@
+
+        #[cfg(feature = "bpf")]
+        {
+            info!("Registering probes: {:?}", found_probes);
```
Move this to debug level

kylebartush-twitter replied:
Done

brayniac commented:
Is it possible to improve the error message when the lib path is incorrect?

```
2021-06-08 19:00:19.097 INFO  [rezolus] Registering probes: [("/usr/lib/x86_64-linux-gnu/libkrb5.so.26", "krb26", "krb5_cc_get_principal"), ("/usr/lib/x86_64-linux-gnu/libkrb5.so.26", "krb26", "krb5_parse_name_flags"), ("/usr/lib/x86_64-linux-gnu/libkrb5.so.3", "krb3", "krb5_cc_get_principal"), ("/usr/lib/x86_64-linux-gnu/libkrb5.so.3", "krb3", "krb5_parse_name_flags"), ("/lib64/libcurl.so.4", "curl", "curl_global_init")]
2021-06-08 19:00:22.108 ERROR [rezolus] failed to initialize usercall sampler
```

This was not super helpful for figuring out which of the probes failed to attach. Would be nice if we could log an error with the library name and that the path is invalid - on this test system, krb libs were under /usr/lib64 - but it wasn't obvious why it failed to initialize until i checked for the libraries manually.

kylebartush-twitter committed:
Change probe registering logs to debug

kylebartush-twitter committed:
Better error handling

- Errors now bubble up more info on why init_bpf failed
- If fault_tolerant is set a single probe configuration error will now
just log an error while keeping the other probes functional

kylebartush-twitter committed:
Code cleanup

- Simplify UsercallStatistic
- Remove unprobed hashmap, instead update statistics

kylebartush-twitter commented:
I've added more detailed errors and made it so a single bad probe doesn't disable the sampler if fault_tolerant is set.

brayniac reviewed (approved):
Looks good to me. 

Ignoring the CI failures on some combinations since that issue does not look to be related to this change.

Manual testing is looking good and the error messages are much better.
